File: app/src/main/java/com/madebyaron/betterhealth/view/list/ListFragment.java
package com.madebyaron.betterhealth.view.list;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.madebyaron.betterhealth.R;
import com.madebyaron.betterhealth.databinding.FragmentListBinding;
import com.madebyaron.betterhealth.view.list.dialogs.AddCaloriesDialog;
import com.madebyaron.betterhealth.view.list.dialogs.AddCigarettesDialog;
import com.madebyaron.betterhealth.view.list.dialogs.AddDrinkDialog;
import com.madebyaron.betterhealth.view.list.dialogs.AddWeightDialog;
import com.madebyaron.betterhealth.viewmodel.ListModel;

public class ListFragment extends Fragment {

    private FragmentListBinding binding;
    private ListAdapter adapter;
    private boolean addButtonExpanded = false;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentListBinding.inflate(inflater, container, false);
        binding.setLifecycleOwner(getViewLifecycleOwner());
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        if (getActivity() != null) {
            adapter = new ListAdapter(getActivity());
            adapter.attachToRecyclerView(binding.listRecyclerView);
            binding.listRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

            binding.listSwipeRefresh.setColorSchemeColors(
                    ContextCompat.getColor(getActivity(), R.color.color_control)
            );
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (getActivity() != null) {
            ListModel model = ViewModelProviders.of(getActivity()).get(ListModel.class);

            binding.setFragment(this);
            binding.setModel(model);

            model.dataList.observe(getViewLifecycleOwner(), adapter::setData);
            model.loading.observe(getViewLifecycleOwner(), binding.listSwipeRefresh::setRefreshing);
            binding.listSwipeRefresh.setOnRefreshListener(model::reload);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        compressAddButton();
    }

    public void add() {
        if (addButtonExpanded) {
            compressAddButton();
        } else {
            expandAddButton();
        }
    }

    public void addWeight() {
        compressAddButton();
        if (getActivity() != null) new AddWeightDialog(getActivity()).show();
    }

    public void addCalories() {
        compressAddButton();
        if (getActivity() != null) new AddCaloriesDialog(getActivity()).show();
    }

    public void addDrink() {
        compressAddButton();
        if (getActivity() != null) new AddDrinkDialog(getActivity()).show();
    }

    public void addCigarettes() {
        compressAddButton();
        if (getActivity() != null) new AddCigarettesDialog(getActivity()).show();
    }

    private void compressAddButton() {
        addButtonExpanded = false;

        binding.listAddWeightButton.hide();
        binding.listAddCaloriesButton.hide();
        binding.listAddDrinkButton.hide();
        binding.listAddCigarettesButton.hide();

        binding.listAddButton.startAnimation(
                AnimationUtils.loadAnimation(
                        getContext(),
                        R.anim.compress_floating_action_button
                )
        );
    }

    private void expandAddButton() {
        addButtonExpanded = true;

        binding.listAddWeightButton.show();
        binding.listAddCaloriesButton.show();
        binding.listAddDrinkButton.show();
        binding.listAddCigarettesButton.show();

        binding.listAddButton.startAnimation(
                AnimationUtils.loadAnimation(
                        getContext(),
                        R.anim.expand_floating_action_button
                )
        );
    }
}

File: app/src/main/java/com/madebyaron/betterhealth/view/list/ListFragment.kt
package com.madebyaron.betterhealth.view.list

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.madebyaron.betterhealth.R
import com.madebyaron.betterhealth.data.Data
import com.madebyaron.betterhealth.databinding.FragmentListBinding
import com.madebyaron.betterhealth.view.list.dialogs.AddCaloriesDialog
import com.madebyaron.betterhealth.view.list.dialogs.AddCigarettesDialog
import com.madebyaron.betterhealth.view.list.dialogs.AddDrinkDialog
import com.madebyaron.betterhealth.view.list.dialogs.AddWeightDialog
import com.madebyaron.betterhealth.viewmodel.ListModel
import kotlinx.android.synthetic.main.fragment_list.*

class ListFragment : Fragment() {

	private lateinit var binding: FragmentListBinding
	private lateinit var adapter: ListAdapter
	private var addButtonExpanded = false

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
		binding = FragmentListBinding.inflate(inflater, container, false)
		binding.lifecycleOwner = viewLifecycleOwner
		return binding.root
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		activity?.let {
			adapter = ListAdapter(it)
			adapter.attachToRecyclerView(list_recycler_view)
			list_recycler_view.layoutManager = LinearLayoutManager(context)

			list_swipe_refresh.setColorSchemeColors(
					ContextCompat.getColor(it, R.color.color_control)
			)
		}
	}

	override fun onActivityCreated(savedInstanceState: Bundle?) {
		super.onActivityCreated(savedInstanceState)
		activity?.let {
			val model = ViewModelProviders.of(it).get(ListModel::class.java)

			binding.fragment = this
			binding.model = model

			model.dataList.observe(viewLifecycleOwner, Observer<List<Data>> { dataList ->
				adapter.dataList = dataList
			})
			model.loading.observe(viewLifecycleOwner, Observer<Boolean> { loading ->
				list_swipe_refresh.isRefreshing = loading
			})
			list_swipe_refresh.setOnRefreshListener {
				model.reload()
			}
		}
	}

	override fun onPause() {
		super.onPause()
		compressAddButton()
	}

	fun add() {
		if (addButtonExpanded) {
			compressAddButton()
		} else {
			expandAddButton()
		}
	}

	fun addWeight() {
		compressAddButton()
		activity?.let { activity -> AddWeightDialog(activity).show() }
	}

	fun addCalories() {
		compressAddButton()
		activity?.let { activity -> AddCaloriesDialog(activity).show() }
	}

	fun addDrink() {
		compressAddButton()
		activity?.let { activity -> AddDrinkDialog(activity).show() }
	}

	fun addCigarettes() {
		compressAddButton()
		activity?.let { activity -> AddCigarettesDialog(activity).show() }
	}

	private fun compressAddButton() {
		addButtonExpanded = false

		list_add_weight_button.hide()
		list_add_calories_button.hide()
		list_add_drink_button.hide()
		list_add_cigarettes_button.hide()

		list_add_button.startAnimation(
				AnimationUtils.loadAnimation(
						context,
						R.anim.compress_floating_action_button
				)
		)
	}

	private fun expandAddButton() {
		addButtonExpanded = true

		list_add_weight_button.show()
		list_add_calories_button.show()
		list_add_drink_button.show()
		list_add_cigarettes_button.show()

		list_add_button.startAnimation(
				AnimationUtils.loadAnimation(
						context,
						R.anim.expand_floating_action_button
				)
		)
	}
}